boards: add support for Qorvo DWM3001CDK

ChitlangeSahas opened the pull request:
- creates the DWM3001C DK board
- Verified that building and flashing of `blinky` sample worked.

ChitlangeSahas committed:
boards: support for DWM3001CDK

Board link: https://www.qorvo.com/products/p/DWM3001CDK

Signed-off-by: Sahas Chitlange <chitlangesahas@gmail.com>

ChitlangeSahas commented on boards/arm/decawave_dwm3001c_dev/decawave_dwm3001c_dev.dts:
```diff
@@ -0,0 +1,165 @@
+	pinctrl-0 = <&spi3_default>;
+	pinctrl-1 = <&spi3_sleep>;
+	pinctrl-names = "default", "sleep";
+	ieee802154alias: dw3001c@0 {
```
`devicetree error: Label 'ieee802154' appears on /soc/radio@40001000/ieee802154 and on /soc/spi@4002f000/dw3001c@0`

Kept getting that error so I changed the label name.

fgrandel replied:
The nRF52833 SoC (DWM300x) has a ieee802154 device node in its base definition while the nRF52832 (DWM100x) has not. So the ieee802154 node name is already taken (see https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/nordic/nrf52833.dtsi#L95).

Now there's [only 2 hard things in computer science](https://martinfowler.com/bliki/TwoHardThings.html) - and one of them is naming:

Boards rarely expose more than one ieee802154 capable radio which is also the reason why most drivers are not yet instance capable so we've gotten away with the ieee802154 name and single instance drivers for quite some time. But this is going to change with ranging.

In the general case, I see a few reasons why s/o would have several ieee802154 instances on a single board:
1) PHY diversity (BLE + UWB)
2) PDoA/AoA (one transceiver/driver instance per antenna if the transceiver is not multi-antenna capable - see the qorvo transceivers that cater for such a case)
3) Participating in several PANs on different channels at the same time. The standard explicitly proposes this scenario but we don't have such a requirement from the community, yet. 

[Devicetree](https://github.com/devicetree-org/devicetree-specification/releases/tag/v0.4) uses a `peripheral-name: hardware-name@memory-address` naming convention. The hardware name may be repeated (you can have several instances of the same hardware at different addresses), the peripheral name cannot as it is instance specific.

The peripheral name is usually the one from the datasheet. But not so for the ieee802154 node as it designates a sub-node of a multi-driver capable peripheral or bus.

We have one multi-driver precedence for the CC13/26xx: ieee802154 (2.4G) and ieee802154g (SubG). I don't like the one-letter postfix very much, though. It's neither scalable nor readable.

Therefore I propose the following naming scheme:
- ieee802154 for the main data communication device on a board
- ieee802154-<phy> for devices that support additional PHYs on the same board (e.g. for ranging or when translating/routing between PHYs)
- ieee802154[-<phy>]-<n> for boards with several driver instances of the same PHY for PDoA/AoA or channel diversity (n zero-based integers)

The PHY-suffix must be chosen "creatively" per board (e.g. "subg" or "sun" rather than "sun-fsk") - as it may not always cleanly map to a single PHY when a driver supports several of them. But I think we should keep it pragmatic as we'll probably always find some semantically matching string in the datasheet of a given transceiver.

@rlubos, @jukkar, @nandojve Would love to hear your opinion on those thoughts if you find the time.

ChitlangeSahas replied:
Changed the name to `ieee802154_uwb` `ieee802154-uwb` seems to throw errors, is hypen not valid?

ChitlangeSahas commented on boards/arm/decawave_dwm3001c_dev/Kconfig:
```diff
@@ -0,0 +1,10 @@
+config BOARD_ENABLE_DCDC
+	bool "DCDC mode"
+	select SOC_DCDC_NRF53X_APP # todo(sahas): is this right?
+	default n
```
If left `y` I get this:

``` 

warning: SOC_DCDC_NRF53X_APP (defined at soc/arm/nordic_nrf/nrf53/Kconfig.soc:79) has direct dependencies SOC_NRF5340_CPUAPP && SOC_SERIES_NRF53X && SOC_FAMILY_NRF with value n, but is currently being y-selected by the following symbols:
 - BOARD_ENABLE_DCDC (defined at boards/arm/decawave_dwm3001c_dev/Kconfig:6), with value y, direct dependencies BOARD_DECAWAVE_DWM3001C_DEV (value: y), and select condition BOARD_DECAWAVE_DWM3001C_DEV (value: y)
Parsing /Users/sahaschitlange/Desktop/FindMyCat-org/zephyrproject/zephyr/Kconfig
Loaded configuration '/Users/sahaschitlange/Desktop/FindMyCat-org/zephyrproject/zephyr/boards/arm/decawave_dwm3001c_dev/decawave_dwm3001c_dev_defconfig'
Merged configuration '/Users/sahaschitlange/Desktop/FindMyCat-org/zephyrproject/zephyr/samples/basic/blinky/prj.conf'

error: Aborting due to Kconfig warnings
```

fgrandel commented:
Please review the [Zephyr Contribution Guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html) and [Zephyr Coding Guidelines](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html) specifically the [Commit Message Guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines). Your commit title scheme does not yet match.

You should prefix with `drivers: ieee802154: dw1000` while we haven't renamed the driver `drivers: ieee802154: qorvo` after the rename. Similarly for board or subsystem changes. Just look through git logs for precedences and examples.

fgrandel reviewed (commented):
Thank you very much for your work - this is really advancing well!

The main problem with this patch set from a (preliminary) reviewer perspective is that it doesn't present your changes in self-contained commits, yet. You need to use `git rebase -i` to fix this. Maybe you're even aware of that - after all this is just a draft (please disregard all my comments regarding commit squashing in that case).

I'd nevertheless propose you fix your commit history before my next (preliminary) review round, just to make it a little easier for me to support you.

fgrandel reviewed (dismissed):
Changes requested re commit titles/structure and comments indicated as such in the review.

ChitlangeSahas commented on boards/arm/decawave_dwm3001c_dev/Kconfig:
```diff
@@ -0,0 +1,10 @@
+# DecaWave DWM3001C DEV board configuration
```
Renamed all files to qorvo; changed board name to DWM3001CDK.

ChitlangeSahas commented on boards/arm/decawave_dwm3001c_dev/Kconfig:
```diff
@@ -0,0 +1,10 @@
+
+config BOARD_ENABLE_DCDC
+	bool "DCDC mode"
+	select SOC_DCDC_NRF53X_APP # todo(sahas): is this right?
```
fixed ✅

ChitlangeSahas replied:
Just realized; the reason I thought it was NRF53X was because somehow my brain concluded nrf52833 was nrf53 series xD; all good; sorry for the confusion.

ChitlangeSahas commented on boards/arm/decawave_dwm3001c_dev/Kconfig:
```diff
@@ -0,0 +1,10 @@
+# DecaWave DWM3001C DEV board configuration
+
+# Copyright (c) 2019 Stéphane D'Alu
```
Added `, Zephyr Project` to the copyright information in files I created.

ChitlangeSahas commented on boards/arm/decawave_dwm3001c_dev/doc/index.rst:
```diff
@@ -0,0 +1,54 @@
+********
+
+The DWM3001CDK development board includes the DWM3001C module, battery
+connector and charging circuit, LEDs, buttons, Raspberry-Pi and USB
```
✅

ChitlangeSahas commented on boards/arm/decawave_dwm3001c_dev/decawave_dwm3001c_dev-pinctrl.dtsi:
```diff
@@ -34,41 +34,9 @@
-			low-power-enable;
-		};
-	};
+	// Todo(Sahas): Potentially add SPI1 and SPI2;
```
Fixed. 👍

ChitlangeSahas committed:
dts: bindings: ieee802154: add qorvo dwm3001c bindings

Adds the bindings for DWM3001C UWB transceiver.

Signed-off-by: Sahas Chitlange <chitlangesahas@gmail.com>

ChitlangeSahas commented on boards/arm/decawave_dwm3001c_dev/decawave_dwm3001c_dev.dts:
```diff
@@ -74,7 +66,6 @@
 		led1-red   = &led1_red;
 		led2-blue  = &led2_blue;
 		led3-green = &led3_green;
-		// pwm-led0   = &pwm_led0_red;
```
Great point; I added the PWM stuff back.

ChitlangeSahas committed:
dts: bindings: ieee802154: add todo for moving antenna rx/tx properties

The properties should exist in ieee802154 drivers instead; creating a todo in dwm3001cdk and dwm1001-dev boards.

Signed-off-by: Sahas Chitlange <chitlangesahas@gmail.com>

ChitlangeSahas commented on dts/bindings/ieee802154/decawave,dw3001c.yaml:
```diff
@@ -0,0 +1,35 @@
+      If connected directly the MCU pin should be configured
+      as active low.
+
+  tx-ant-delay:
```
Done ✅

ChitlangeSahas commented on boards/arm/decawave_dwm3001c_dev/decawave_dwm3001c_dev.dts:
```diff
@@ -0,0 +1,186 @@
+
+&uart0 {
+	status = "okay";
+	compatible = "nordic,nrf-uart";
```
I think I got them all.

ChitlangeSahas commented on boards/arm/decawave_dwm3001c_dev/decawave_dwm3001c_dev.dts:
```diff
@@ -0,0 +1,186 @@
+
+&flash0 {
+
+	partitions {
```
Yeah after your suggestion I looked through some; it seems similar.

ChitlangeSahas commented on boards/arm/decawave_dwm3001c_dev/decawave_dwm3001c_dev.dts:
```diff
@@ -0,0 +1,186 @@
+		zephyr,sram           = &sram0;
+		zephyr,flash          = &flash0;
+		zephyr,code-partition = &slot0_partition;
+		zephyr,ieee802154     = &ieee802154;
```
Done!

fgrandel reviewed (commented):
Hi Sahas!

All my previous comments have been resolved. Nice work!

You have a few failing compliance checks - but see for yourself. Probably your commit hook does not yet work correctly otherwise it would have catched this.

Before the board can be merged we have to build up some basic driver support for the DW300x of course. The correct ordering of changes (commits) will place the driver support before (!) the board support: The driver improvements can be merged w/o the board but not the other way round.

You can easily achieve this with `git rebase -i` as we discussed. Just build up your commits on top of the existing changes. You can move them backwards later with the "pick" command in interactive rebase. You won't get conflicts as the files being changed are different from the files you added so far.

Please leave the PR in "draft" status while you work on the driver. We might have to split it up in several PRs later for merging but for the moment being I'd say it's fine to collect and continuously review your changes in this PR.

If you ensure that you push rebases on a different main version separately from actual changes to your commits I can easily review each delta. Try the "compare" button besides every push recorded in the PR yourself so you get a feeling for it.

ChitlangeSahas committed:
chore: add initial files

ChitlangeSahas committed:
drivers: ieee802154: separate common methods from dw1000 driver

jukkar commented:
@ChitlangeSahas are you continuing with this PR, there are some compliance check failures that should be easy to fix.

nordicjm reviewed (changes requested):
Needs to be migrated to hwmv2
